File: bin/command/impl/build_main_image.py
import os

from command.command import Command
from command.command_result import CommandResult
from context.impl.run_chain_context import RunChainContext
from util import utils, docker_utils


class BuildMainImage(Command):
    """Builds the main image."""

    context: RunChainContext

    def __init__(self, context: RunChainContext) -> None:
        self.context = context

    def execute(self) -> CommandResult:
        main_image = self.context.config['DOCKER']['main_image']

        if not self.context.args['--rebuild'] and docker_utils.item_exists('image', main_image):
            utils.warn(f"Image '{main_image}' already exists, not building")
            return CommandResult.SKIPPED

        utils.log(f"Building '{main_image}' image")
        main_image_cmd = [
            'docker',
            'build',
            '--tag',
            main_image,
            '--file',
            os.path.join('docker', 'Dockerfile'),
            '.',
        ]

        if self.context.args['--no-cache']:
            main_image_cmd.insert(2, '--no-cache')
        else:
            for item in self.context.args['--cache-from']:
                main_image_cmd[2:2] = ['--cache-from', item]

        if self.context.args['--suspend'] or self.context.args['--debug']:
            main_image_cmd[2:2] = ['--build-arg', 'suspend=true' if self.context.args['--suspend'] else 'debug=true']

        completed_process = utils.execute_cmd(main_image_cmd)
        if 0 != completed_process.returncode:
            utils.warn(f"Failed running the following command: {main_image_cmd}")
            return CommandResult.FAILED

        return CommandResult.OK

File: runner/src/main/java/dev/amrw/runner/chain/run/command/BuildMainImage.java
package dev.amrw.runner.chain.run.command;

import dev.amrw.runner.callback.BuildImageResultCallback;
import lombok.extern.log4j.Log4j2;
import org.apache.commons.chain.Command;
import org.apache.commons.chain.Context;

import java.io.File;
import java.util.Set;
import java.util.stream.Collectors;

/**
 * Builds the build image.
 */
@Log4j2
public class BuildMainImage extends RunChainCommand {

    @Override
    public boolean execute(final Context context) {
        super.prepareContext(context);
        final var mainImageName = dockerConfig.getMainImageConfig().getName();

        if (!args.rebuild() && runChainContext.mainImageExists()) {
            log.info("Image already exists, not building (name={})", mainImageName);
            return Command.CONTINUE_PROCESSING;
        }

        if (args.noCache()) {
            // When the `--no-cache` Docker option for the `build` command has been specified, the existing image with
            // the same tag is going to be re-tagged to `<none>:<none>` and left behind (it'll be dangling). This is
            // why it's best to remove the existing images before building from scratch as to not leave an untagged
            // image behind.
            final var images = dockerClientHelper.findImagesByName(mainImageName);
            images.forEach(image -> {
                log.debug("Removing image (repoTags={}, id={})", image.getRepoTags(), image.getId());
                dockerClient.removeImageCmd(image.getId()).exec();
            });
        }

        buildImage(mainImageName);

        return Command.CONTINUE_PROCESSING;
    }

    private void buildImage(final String imageName) {
        log.info("Building image (name={})", imageName);
        final var baseDirectory = new File(dockerConfig.getBaseDirPath());
        log.trace("Docker base directory: {}", baseDirectory);
        final var dockerfileMain = new File(dockerConfig.getDockerfileMainPath());
        log.trace("Dockerfile: {}", dockerfileMain);
        final var mainImageCmd = dockerClient.buildImageCmd()
                // NOTE: There's a dependency on `baseDirectory` inside `withDockerfile()`, therefore
                // `withBaseDirectory()` _must_ come before `withDockerfile()`, if applicable. Otherwise, this error
                // may be thrown:
                // > Dockerfile is excluded by pattern '*' in .dockerignore file
                .withTags(Set.of(imageName))
                .withBuildArg("debug", Boolean.toString(args.debug()))
                .withBuildArg("suspend", Boolean.toString(args.suspend()))
                .withBaseDirectory(baseDirectory)
                .withDockerfile(dockerfileMain);

        if (args.noCache()) {
            mainImageCmd.withNoCache(true);
        } else {
            log.trace("Caching from:\n{}", args.cacheFrom());
            mainImageCmd.withCacheFrom(args.cacheFrom().stream().collect(Collectors.toUnmodifiableSet()));
        }

        final var imageId = mainImageCmd.exec(new BuildImageResultCallback()).awaitImageId();
        log.debug("Created image (name={}, id={})", imageName, imageId);
    }
}
